### ISSUE TITLE
[RFC] PVC recycle should wait for pod termination.

### ISSUE THREAD

**Issue by swagiaal** [kubernetes/kubernetes#25101]
I know we have talked about this before but I wanted to get some discussion on the record as I think we discussed this in the context of protecting the binder state as opposed to user protection. I ran into a use case which exposes this problem.

The https://github.com/openshift/origin-metrics has a redeploy option which does a delete  of all pods and PVCs and recreates them. The issue is that the cassandra pod takes a while to shut down meanwhile the deletion of the PVC triggers recycling.

For NFS based volumes this creates a problem; when a file that is still open is deleted the NFS client renames it to .nfsXYZ (http://nfs.sourceforge.net/#faq_d2). The recycler sees those files and is unable to delete the directory.

I am not sure if there is a mechanism to do this but one possible solution would be to add the PVC to a list of things which are deleted when a pod is terminated.

**Comment by jsafrane** [comment]
I got bitten by this behavior in deleter e2e tests - Kubernetes tries to delete a PV that is still attached to a node, while kubelet on the node is slowly trying to detach it. Quick brainstorming:

* We can delete/recycle volume when appropriate PVC is deleted *and* the PV is unmounted + detached from all nodes, i.e. we need cooperation between attach&detach controller and recycler.

* Alternatively, I can imagine that Kubernetes refuses to delete PVC that is attached somewhere with relevant 4xx error

* Or, instead of deleting PVC we can mark it as "Terminating" (same as in pods) and delete it only after appropriate volume is detached from all nodes. In this case, we must block all new pods that would like to use the PVC, on the other hand we don't need any changes in recycler./deleter logic.

The most clean solution to me is the first one - let the PVC to be deleted and postpone recycling until it's detached.

**Comment by rootfs** [comment]
@kubernetes/sig-storage 

I think the Jan's first solution is a sound one. But from an end user perspective, it is just odd to the PVC is deleted while it is still be used, an inexperienced/uninformed user may even believe the PV is also deleted as a result. On the other hand, on GCE, it is not permitted to delete a disk while it is still being used by a VM. 

I would prefer Jan's 2nd solution, although this could mean a PVC is more stateful.

**Comment by markturansky** [comment]
Related but closed -- https://github.com/kubernetes/kubernetes/issues/12084

**Comment by bprashanth** [comment]
It is waay too easy to leak pvs today. 
Simply doing: 
```
 $ kubectl create -f  ~/rtmp/volumes/volume_rc.yaml 
persistentvolumeclaim "pv-provisioning-test" created
replicationcontroller "gce-busybox" created
$ kubectl delete rc, po, pvc --all
```
will leak. 

It doesn't sound like we're fixing this for 1.3, but maybe we can simply throw an error when a user tries to delete that pvc?
```
Failed to delete pvc foo, currently being used by pod bar. Use -f to override.
```

**Comment by childsb** [comment]
we believe this is fixed in 1.3.4 .
